Conserve component density

mgjarrett opened the pull request:
## Description

The `axialExpansionChanger` was implemented to conserve the initial mass of each component in a block with axial expansion determined by the target component, using the funciton `_conserveComponentMass`. Since different components are made of different materials that thermally expand at different rates, this necessarily means that the densities have to be adjusted artificially for the mass to be conserved over the block for each individual component.

After testing and discussing the methodology, it has been decided that it is preferred to conserve the realistic density of the material rather than conserving mass within a block. Effectively, we are allowing mass to move between blocks.

The `Tinput` values in the test reactor for `detailedAxialExpansion` were updated because the unit test to ensure density conservation requires that the `Tinput` is equal to the material reference temperature for calculating thermal expansion factors.

---

## Checklist

- [x] This PR has only one purpose or idea.
- [x] Tests have been added/updated to verify that the new/changed code works.

- [x] The code style follows [good practices](https://terrapower.github.io/armi/developer/standards_and_practices.html).
- [x] The commit message(s) follow [good practices](https://terrapower.github.io/armi/developer/tooling.html).

- [x] The [release notes](https://terrapower.github.io/armi/release/index.html) are up-to-date with any bug fixes or new features.
- [x] The documentation is still up-to-date in the `doc` folder.
- [x] The dependencies are still up-to-date in `setup.py`.

mgjarrett commented:
@albeanth I took a quick glance at the docs/release notes and didn't see anything that needed to be updated, but you would probably know better than me.

jakehader commented on armi/reactor/converters/axialExpansionChanger.py:
```diff
@@ -284,24 +286,43 @@ def manageCoreMesh(self, r):
-        list containing block heights pre-expansion
-    oldVolume : list of floats
-        list containing component volumes pre-expansion
+        solidMaterials = _getSolidComponents(b)
```
Should be `solidComponents`.

mgjarrett replied:
Updated

jakehader commented on armi/tests/detailedAxialExpansion/refSmallReactorBase.yaml:
```diff
@@ -160,15 +160,15 @@ blocks:
             shape: Helix
             material: HT9
-            Tinput: 25.0
+            Tinput: 27.548
```
Can you explain in the PR description why these are changing? I assume that this is because we have `inputHeightsConsideredHot` set to `True` for the tests? Not clear though.

mgjarrett replied:
The reference temperature of HT9 for thermal expansion is 27.548C. The test to check tohat density is conserved only works if Tinput is equal to the reference temperature for thermal expanison.

albeanth replied:
In line with @jakehader 's comment:

@mgjarrett - for our future selves I think it would be good to leave a comment in this PR on precisely why you/we are making the changes to these temperatures and to armi/materials/zr.py (especially the latter). 

I could easily see us coming back a few months later and being totally stumped as to why these temperatures were all changed.

mgjarrett replied:
@albeanth are you looking for a comment directly in the yaml file? Or just include it in the commit message?

albeanth replied:
The commit message would be fine, but I was thinking about just a general comment in this PR. That way if someone in the future does a git blame and comes back to this PR it will be (somewhat) front and center.

I suppose a comment in the yaml would be ok too - maybe at the very top right off the bat? Not sure where it's best to include it. I'll leave it to your discretion.

jakehader reviewed (changes requested):
See comments.

albeanth commented:
@mgjarrett Haven't had a chance to review this yet. But am noticing that the coverage dropped pretty drastically. Especially in axialExpansionChanger.py. Can you confirm that you are getting the same coverage drop on your local machine?

mgjarrett commented:
What's confusing is that the coverage rose in commit `ec29050`

The next 3 commits are just changing function and variable names, but no functional changes have been implemented at all. Now it says coverage decreased. One of those commits broke a unit test, which might have confused coveralls.

mgjarrett commented:
@jakehader I've addressed your comments. I'm trying to understand what happened to the coverage but I'm having trouble understanding what happened from the coveralls report. It does appear that all of the new lines are covered.

albeanth commented on armi/reactor/converters/tests/test_axialExpansionChanger.py:
```diff
@@ -810,6 +824,19 @@ def test_coldAssemblyExpansion(self):
+                        for cExp in bExp:
+                            if not isinstance(cExp.material, custom.Custom):
+                                matDens = cExp.material.density3(Tc=cExp.temperatureInC)
+                                compDens = cExp.getMassDensity()
```
`matDens` and `compDens` are unused variables.

albeanth reviewed (approved):
Approving, but would love to see a comment addressing this. Thanks @mgjarrett ! 
https://github.com/terrapower/armi/pull/846#discussion_r951981623

jakehader commented:
Okay, after talking with @mgjarrett I think I understand the scope of this change. Basically, we are demonstrating that _if_ the component's input temperature is equal to its material's `refTemp` then thermal expansion will be correctly treated. Otherwise, if these inputs are not equal then the `material.density3()` and `c.getMassDensity()` will not be equal and therefore be off by a factor of the thermal expansion coefficients between the material at the reference temperature and the input temperatures. Please correct my @mgjarrett if this is wrong.

If the above is correct, then I suggest adding a _warning_ to the user that the component input reference does not align with the material's `refTemp`, which can probably be added here: https://github.com/terrapower/armi/blob/main/armi/reactor/components/component.py#L318-L328

I am not sure, but this will probably break downstream unit tests that we will have to fix. @opotowsky, would you mind pulling this in and running all internal plugin unit tests and the integration tests?

mgjarrett commented:
@jakehader that summary is correct. I've added a warning when the absolute value of the thermal expansion factor at `Tinput` is greater than 1e-6.

I am firing off some downstream tests right now. I expect the impacts to be minimal but we will see.

jakehader commented on armi/reactor/components/component.py:
```diff
@@ -232,11 +232,26 @@ def __init__(
+            self.material.linearExpansionPercent(Tc=self.inputTemperatureInC) / 100.0
+        )
+        if not (abs(expansionFactor) < 1.0e-6):
+            runLog.warning(
```
Let's make this a single warning so it doesnt show up a million times in the stdout

mgjarrett replied:
Done

albeanth commented:
> @albeanth I took a quick glance at the docs/release notes and didn't see anything that needed to be updated, but you would probably know better than me.

To my knowledge, no need to add anything to the docs. 

I would add one line to the release notes commenting on the density conservation bug fix.

mgjarrett commented:
Downstream tests are currently failing; waiting to merge this until we get that resolved

opotowsky commented:
Downstream tests resolved -- merging!

onufer commented:
I would expect that after #820 is resolved the odd initial temps in blueprints would not longer be needed

john-science commented on armi/reactor/components/component.py:
```diff
@@ -232,11 +232,27 @@ def __init__(
         self.p.mergeWith = mergeWith
         self.p.customIsotopicsName = isotopics
 
+    def tInputWarning(self, Tinput):
```
I think we should move this to the unit tests.

Validating the input data during run time makes our code slower. And we can't do ALL of the validation in the world at run time.

For instance, if we add this as a unit test here, we will automatically run this validation against every material in the repo:

https://github.com/terrapower/armi/blob/main/armi/materials/tests/test_materials.py#L28-L34

A benefit of that is we can control where this validation gets done. 

For instance, what if this Component is made of a void / empty material. Do we expect that material to have the `linearExpansionPercent()` method?

I think I will open a ticket to move this to unit tests.

john-science replied:
I'll handle it:  https://github.com/terrapower/armi/issues/861

mgjarrett replied:
1. This is checking the user input, so it's not possible to perform the check as a unit test.
2. #860 removes the code in question because it fixed the root of the problem.
